homeassistant/components/logbook.py

```diff
@@ -317,43 +317,90 @@ def humanify(hass, events):
+    from homeassistant.components.recorder.util import \
+        RETRIES, QUERY_RETRY_WAIT
+    from sqlalchemy.exc import SQLAlchemyError
+    import time
+
+    timer_start = time.perf_counter()
+
+    query = session.query(States).with_entities(States.entity_id).distinct()
+
+    for tryno in range(0, RETRIES):
+        try:
+            result = [
+                row.entity_id for row in query
+                if entity_filter(row.entity_id)]
+
+            if _LOGGER.isEnabledFor(logging.DEBUG):
+                elapsed = time.perf_counter() - timer_start
+                _LOGGER.debug(
+                    'fetching %d distinct domain/entity_id pairs took %fs',
+                    len(result),
+                    elapsed)
+
+            return result
+        except SQLAlchemyError as err:
+            _LOGGER.error("Error executing query: %s", err)
+
+            if tryno == RETRIES - 1:
+                raise
+            else:
+                time.sleep(QUERY_RETRY_WAIT)
+
+
+def _generate_filter_from_config(config):
+    from homeassistant.helpers.entityfilter import generate_filter
+
+    excluded_entities = []
+    excluded_domains = []
+    included_entities = []
+    included_domains = []
+
+    exclude = config.get(CONF_EXCLUDE)
+    if exclude:
+        excluded_entities = exclude.get(CONF_ENTITIES, [])
+        excluded_domains = exclude.get(CONF_DOMAINS, [])
+    include = config.get(CONF_INCLUDE)
+    if include:
+        included_entities = include.get(CONF_ENTITIES, [])
+        included_domains = include.get(CONF_DOMAINS, [])
+
+    return generate_filter(included_domains, included_entities,
+                           excluded_domains, excluded_entities)
+
+
 def _get_events(hass, config, start_day, end_day, entity_id=None):
     """Get events for a period of time."""
     from homeassistant.components.recorder.models import Events, States
     from homeassistant.components.recorder.util import (
         execute, session_scope)
 
+    entities_filter = _generate_filter_from_config(config)
+
     with session_scope(hass=hass) as session:
+        if entity_id is not None:
+            entity_ids = [entity_id.lower()]
+        else:
+            entity_ids = _get_related_entity_ids(session, entities_filter)
+
         query = session.query(Events).order_by(Events.time_fired) \
-            .outerjoin(States, (Events.event_id == States.event_id))  \
+            .outerjoin(States, (Events.event_id == States.event_id)) \
             .filter(Events.event_type.in_(ALL_EVENT_TYPES)) \
             .filter((Events.time_fired > start_day)
                     & (Events.time_fired < end_day)) \
             .filter((States.last_updated == States.last_changed)
-                    | (States.state_id.is_(None)))
-
-        if entity_id is not None:
-            query = query.filter(States.entity_id == entity_id.lower())
+                    | (States.state_id.is_(None))) \
+            .filter(States.entity_id.in_(entity_ids))
```

Comment:
Realized we're doing more work than before. We should only filter stuff if a user has supplied a filter (most won't).

Reply:
Right. Hope its not a big deal as logbook is not really usable on large datasets, and on small ones speed should not be an issue.

I will add additional PR to skip filtering if exclude/include lists are totally empty.

Reply:
I have a bigger plan for logbook. Let me write up an architecture issue.

Reply:
https://github.com/home-assistant/architecture/issues/110

Reply:
Going to revert this PR. This change is filter out any logbook entry that has no entity associated with it, like calls to `logbook.log` service.

Reply:
Fix in https://github.com/home-assistant/home-assistant/pull/18721